[DOC-2113] feat(optimizer): Histogram Stats API Update and Enhancement; (#140)
tigergraph/gsql-docs: modules/gsql-v2/querying/partials/histogram-api.adoc

`POST :14240/gsqlserver/gsql/stats/histogram`

This endpoint computes histograms for a specified attribute of a vertex or edge type.

READ_SCHEMA privilege is required for the graph that the histogram attribute belongs to. When gathering for multiple graphs at once (see Compute Multiple Histograms in one request), READ_SCHEMA is required on ALL of those graphs

==== Parameters
Options can be specified in the header using -H:
`GSQL-Timeout` indicates timeout for the request. In the case of multiple computations, this timeout covers the overall timeout, including all the histogram computations

> `GSQL-Thread-Limit` indicates the thread limit per histogram computation. In the case of multiple computations this thread limit will apply to each histogram computation individually

Reviewer: ```suggestion
`GSQL-Timeout` indicates timeout for the request. In the case of multiple computations, this timeout covers the overall timeout, including all the histogram computations.
`GSQL-Thread-Limit` indicates the thread limit per histogram computation. In the case of multiple computations this thread limit will apply to each histogram computation individually.
```

|`bucket`
|The number of intervals (https://en.wikipedia.org/wiki/Data_binning[buckets]) to divide the range of attribute values by.

> By default we use a bucket value of 256. With discrete datatypes, a bucket count that is lower than the specified target may be necessary depending on the data the histogram is computed over.

Reviewer: What do you mean by "specified target"? 
Did you mean "default value"?

Author: By specified target, I am trying to explain how, if we are using say an integer data type with the following values:
```
0, 1, 2, 3
```
If the user suggests a target bucket number of say 10, we will instead use 4. So the user specified target bucket count is a target bucket count, but we may reduce that if it is simply not compatible to the maximum, sensible bucket number (in the example, 4)

----
$ curl --user tigergraph:tigergraph \
-X POST "http://localhost:14240/gsqlserver/gsql/stats/histogram" \
-d '{"graph":"test_graph","vertex":"Comment","attribute":"length","bucket":10}'
----
--

Results::
+
--
[source,console]
----
{
  "error": false,
  "message": "Histogram stats gathering completed",
  "results": [
    {
      "success": "[1/1]",
      "failure": []
    }
  ]
}
----
--
====

The following request computes the histogram data for attribute `joinDate` from edge type `COMP_MEMBER` between from vertex `compForum` and to vertex `compPerson` on graph `test_graph`:

[tabs]
====
Histogram Computation Request::
+
--
.Edge Attribute Histogram Computation
[source.wrap,console]
----
$ curl --user tigergraph:tigergraph \
-X POST "http://localhost:14240/gsqlserver/gsql/stats/histogram" \
-d '{"graph":"test_graph","edge":"COMP_MEMBER","from":"compForum","to":"compPerson","attribute":"joinDate","bucket":10}'
----
--

Results::
+
--
[source,console]
----
{
  "error": false,
  "message": "Histogram stats gathering completed",
  "results": [
    {
      "success": "[1/1]",
      "failure": []
    }
  ]
}
----
--
====

==== Compute Multiple Histograms together

> In addition to individually requesting histogram computation through the POST endpoint, it is also possible to gather ALL histograms on a per schema object basis. `bucket` parameter can still be specified and will be applied to every histogram being computed in the request

Reviewer: ```suggestion
In addition to requesting individual histograms through the POST endpoint, it is also possible to request all valid histograms on a per schema object basis.  The`bucket` parameter can still be specified and will be applied to every histogram being computed in the request.
```

-d '{"bucket":"10"}'
// Compute Histograms for ALL valid attributes under:
// graph test_graph 
// (default bucket value)
-d '{"graph":"test_graph"}'
// Compute Histograms for ALL valid attributes under:
// graph test_graph
// vertex Comment
// (default bucket value)
-d '{"graph":"test_graph", "vertex":"Comment"}'
// Compute Histograms for ALL valid attributes under:
// graph test_graph
// vertex Comment
// (default bucket value)
-d '{"graph":"test_graph", "vertex":"Comment"}'
// Compute Histograms for ALL valid attributes and from/to vertices under:
// graph test_graph
// edge COMP_MEMBER
// (default bucket value)
-d '{"graph":"test_graph", "edge":"COMP_MEMBER"}'
// Compute Histograms for ALL from/to vertices for attribute joinDate:
// graph test_graph
// edge COMP_MEMBER
// attribute joinDate
// (default bucket value)
-d '{"graph":"test_graph", "edge":"COMP_MEMBER", "attribute":"joinDate"}'
// Compute Histograms for ALL valid attributes under:
// graph test_graph
// edge COMP_MEMBER, from vertex compForum, to vertex compPerson
// (default bucket value)
-d '{"graph":"test_graph", "edge":"COMP_MEMBER", "from":"compForum", "to":"compPerson"}'
----

> The output generated is in the same format as individual computation, with the `success` field tracking the total number of successful computations and `failure` will track all of the failures encountered:

Reviewer: ```suggestion
The output generated is in the same format as individual computation, with the `success` field tracking the total number of successful computations and `failure` tracking all of the failures encountered:
```

[tabs]
====
Histogram Computation Request::
+
--
.Gather ALL Histogram Computation
[source.wrap,console]
----
$ curl --user tigergraph:tigergraph \
-X POST "http://localhost:14240/gsqlserver/gsql/stats/histogram" \
-d '{}'
----
--

Results::
+
--
[source,console]
----
{
  "error": false,
  "message": "Histogram stats gathering completed",
  "results": [
    {
      "failure": [],
      "success": "[94/94]"
    }
  ]
}
----
--
====

A more condensed error message will be provided under `failure` if any of the histogram computations encounter an error. This allows the request to log the error and continue with the next computation. For more information on the specific error refer to the logs or run the particular histogram computation individually for more verbose feedback. 

=== Retrieve histogram statistics

> `READ_DATA` privilege is required to retrieve a histogram for a particular attribute. This can be at the attribute level or a higher level. Note that for edge based histograms, the 'from' and 'to' vertex are not read and therefor only the edge requires the privilege

Reviewer: ```suggestion
`READ_DATA` privilege is required to retrieve a histogram for a particular attribute. This can be at the attribute level or a higher level. Note that for edge-based histograms, the 'from' and 'to' vertex are not read and therefore only the edge requires the privilege.
```

[source.wrap,console]
----
$ curl --user tigergraph:tigergraph \
-X DELETE "http://localhost:14240/gsqlserver/gsql/stats/histogram" \
-d '{"graph":"test_graph","edge":"COMP_MEMBER","from":"compForum","to":"compPerson","attribute":"joinDate"}'
----
--

Results::
+
--
[source,console]
----
{
  "error": false,
  "message": "",
  "results": "Deleting any histogram(s) for test_graph.compForum.COMP_MEMBER.compPerson.joinDate"
}
----
--
====

==== Deleting Multiple Histograms together

It is possible to request multiple histograms be deleted in the same request. The graph name is always required, but we may request to delete all histograms under a given type as follows:

[.wrap,console]
----
// Delete Histograms for ALL attributes under:
// graph test_graph
-d '{"graph":"test_graph"}'
// Delete Histograms for ALL attributes under:
// graph test_graph
// vertex Comment
-d '{"graph":"test_graph", "vertex":"Comment"}'
// Delete Histograms for ALL attributes under:
// graph test_graph
// edge COMP_MEMBER, from vertex compForum, to vertex compPerson
-d '{"graph":"test_graph", "edge":"COMP_MEMBER", "from":"compForum", "to":"compPerson"}'
----

==== Current Limitations

> If leader switch or some other interruption takes place in GSQL we will not be able to resume progress. Statistics are stored as they are computed along with their start and end timestamps so there is no progress lost. See recommendations for some suggestions on how to handle such situations

Reviewer: ```suggestion
If leader switch or some other interruption takes place in GSQL the histogram operation will not be able to resume. Statistics are stored as they are computed along with their start and end timestamps so there is no progress lost. See recommendations for some suggestions on how to handle such situations.
```

==== Recommendations

> 1. If timeout or some other event interrupts the collection of all the desired statistics, specifying a smaller subset of histograms based on what still has yet to be gathered will reduce the amount of regathering. For example, if executing a histogram computation request for attributes under all graphs, and after 2/4 graphs are fully gathered we are unable to continue, instead of regathering on all graphs again we can issue two requests at the graph level for graphs 3 and 4

Reviewer: ```suggestion
1. If timeout or some other event interrupts the collection of all the desired statistics, specifying a smaller subset of histograms based on what still has yet to be gathered will reduce the amount of regathering.
For example, if executing a histogram computation request for attributes under all graphs, and after 2/4 graphs are fully gathered a timeout occurs, you can issue two requests at the graph level for graphs 3 and 4.
```

> 2. When possible, it is best to schedule histogram computations after schema change jobs and data loading, not before. This is to ensure the statistics best reflect the current state of the database
> 
> 3. When choosing the right bucket value, the first step is to determine the requirements of the use case. Then, consider the granularity and compaction desired based on accuracy and performance / storage consumption 

Reviewer: ```suggestion
2. When possible, it is best to schedule histogram computations after schema change jobs and data loading, not before. This is to ensure the statistics best reflect the current state of the database.

3. When choosing the right bucket value, the first step is to determine the requirements of the use case. Then, choose the fewer number of buckets that provides enough granularity.
```
